src/observations/topocentric.jl

```diff
@@ -1,151 +1,318 @@
+- `utc::Int`: hours from UTC.
+"""
+@auto_hash_equals struct TimeOfDay
+    light::Symbol
+    start::DateTime
+    stop::DateTime
+    utc::Int 
+    function TimeOfDay(date::DateTime, observatory::ObservatoryMPC{T}) where {T <: AbstractFloat}
+        if issatellite(observatory)
+            return new(:space, date, date, 0)
+        end
+        # Hours from UTC
+        utc = hours_from_UTC(observatory)
+        # Today's sunrise / sunset
+        today = sunriseset(date, observatory)
+        # Yesterday's sunrise / sunset 
+        yesterday = sunriseset(date - Day(1), observatory)
+        # Tomorrow's sunrise / sunset
+        tomorrow = sunriseset(date + Day(1), observatory)
+        # Selection
+        if yesterday[2] <= date <= today[1]
+            return new(:night, Date(yesterday[2]), Date(today[1]), utc)
+        elseif today[1] <= date <= today[2]
+            return new(:day, Date(today[1]), Date(today[2]), utc)
+        elseif today[2] <= date <= tomorrow[1]
+            return new(:night, Date(today[2]), Date(tomorrow[1]), utc)
+        end
+
+    end
+end
+
+TimeOfDay(radec::RadecMPC{T}) where {T <: AbstractFloat} = TimeOfDay(date(radec), observatory(radec))
+
+isday(x::TimeOfDay) = x.light == :day
+isnight(x::TimeOfDay) = x.light == :night
+
+# Print method for TimeOfDay
+# Examples:
+# Night from 2023-06-29 to 2023-06-29 at UTC-7
+# Night from 2023-06-29 to 2023-06-30 at UTC+3
+function show(io::IO, m::TimeOfDay)
+    print(io, uppercasefirst(string(m.light)), " from ", Date(m.start), " to ", Date(m.stop),
+          " at UTC", @sprintf("%+d", m.utc))
+end
+
+@doc raw"""
+    hours_from_UTC(lon::T) where {T <: Real}
+
+Return the naive hour difference between longitude `lon` [rad] and UTC.
+"""
+hours_from_UTC(lon::T) where {T <: Real} = ceil(Int, 12*lon/π - 0.5)
+function hours_from_UTC(observatory::ObservatoryMPC{T}) where  {T <: AbstractFloat} 
+    lon, _ = lonlat(observatory)
+    return hours_from_UTC(lon)
+end
+
+@doc raw"""
+    lonlat(observatory::ObservatoryMPC{T}) where {T <: AbstractFloat}
+
+Return longitude and latitude (both in rad) of an observatory.
+"""
+function lonlat(observatory::ObservatoryMPC{T}) where {T <: AbstractFloat}
+    # ECEF [km]
+    p_ECEF = obsposECEF(observatory)
+    # ECEF [m] -> Geodetic [m]
+    lat_geodetic, lon, altitude = ecef_to_geodetic(1_000 * p_ECEF)
+    # Geodetic [m] -> Geocentric [m]
+    lat_geocentric, _ = geodetic_to_geocentric(lat_geodetic, altitude)
+
+    return lon, lat_geocentric
+end 
+
+@doc raw"""
+    sunriseset(date::DateTime, observatory::ObservatoryMPC{T}) where {T <: AbstractFloat}
+    sunriseset(radec::RadecMPC{T}) where {T <: AbstractFloat}
+
+Return `DateTime` of sunrise and sunset at a particular date and location.
+
+!!! reference
+    See "General Solar Position Calculations" by NOAA at
+    https://gml.noaa.gov/grad/solcalc/solareqns.PDF.
+"""
+function sunriseset(date::DateTime, observatory::ObservatoryMPC{T}) where {T <: AbstractFloat}

```

Comment:
We have the observer's position in ECEF frame as given from the MPC data, and we also know the Sun's ephemeris which we can transform to ECEF frame. Then, it's straight-forward to compute the elevation angle as a function of time, and from there it's straight-forward to compute the sunrise and sunset via root-finding, including possibly the offset due to the solar disk. I think we can get a more accurate computation from there?

Reply:
That's a great suggestion and most likely would improve the accuracy of sunrise(set) times. However, this function only purpose is to group observations by night and that division is usually very neat, so we don't really need a lot of precision in that calculation.  I am in favor of leaving this as it is unless we find a specific asteroid where such a precision makes a difference in the way we group observations.

Reply:
I'm wondering what happens if there's some observations near the local sunrise/sunset, but miss the grouping by, say, one minute, the it gets grouped in the next (or previous) day? Anyway, since this more an accuracy enhancement for edge cases rather than an fix, then we can probably go ahead with the method you propose, and keeping in mind an improvement if it's needed after merging this PR.

Reply:
I've been testing the OD routines with a 5,000 NEOs sample. I will check if such scenario occurs in the sample. If that is the case, we can change to a more accurate method in a future PR.